rfcs/0000-error-handling.md

```diff
@@ -0,0 +1,144 @@
+
+Less catastrophic failures do have a downside: It's harder to notice that anything is wrong at all. We'll include an error in the graphql response for failed subgraphs. We will improve and document the indexing status API, so that developers can programmatically check details on a subgraph's health and build their own alerts and dashboards.
+
+## Urgency
+
+Sooner is better than later.
+
+## Terminology
+
+No new terminology, but the semantics of the "Failed" status are altered.
+
+## Detailed Design
+
+### Error handling
+
+If a WASM handler terminates with a trap, that is an error. The node discards all entity operations and data source operations done by the handler and continue syncing.
+
+For subgraphs that are not yet synced, any handler error halts syncing, so errors in pending versions will not make their way to the current version. 
+
+When the node starts, we currently reset the failed flag on subgraphs. The flag is no longer be reset on synced subgraphs, since the subgraph will already be past the errors.
+
+Any graphql queries to a failed subgraph will include an error of the form:
+
+```
+{
+	message: "Subgraph failed"
+}
+```
+
+So that applications will always be aware that queries may have inconsistent data. Developers should refer to the indexing status API for details on the errors.
+
+### Indexing Status API
+
+The indexing status API will be revamped.
+
+The subgraph indexing status API currently has the top level query:
+
+    indexingStatusesForSubgraphName(subgraphName: String!): [SubgraphIndexingStatus!]!
+
+This is changed to:
+
+    indexingStatusesForSubgraphName(subgraphName: String!): SubgraphIndexingStatus
+
+Which returns null if the name is not found, and returns the status for the current version of the subgraph otherwise.
+
+A field `failedAfter`  is added to record the block previous to the first error.
+
+Errors are made more than just a string with the `Error` type.
+
+The complete indexing status API schema is:
+
+```graphql
+type Query {
+  indexingStatusesForSubgraphName(subgraphName: String!): SubgraphIndexingStatus
+  indexingStatuses(subgraphs: [String!]): [SubgraphIndexingStatus!]!
+}
+
+type SubgraphIndexingStatus {
+  subgraph: String!
+  synced: Boolean!
+  errors: [Error!]! # Sorted from first to last
+  chains: [ChainIndexingStatus!]!
+  node: String!
+}
+
+# Fields are moved from `EthereumIndexingStatus` to the interface.
+interface ChainIndexingStatus {
+  network: String!
+  chainHeadBlock: Block
+  earliestBlock: Block
+  latestBlock: Block
+  failedAfter: Block
+}
+
+type EthereumIndexingStatus implements ChainIndexingStatus { }
+
+# Renamed from `EthereumBlock` to `Block`
+type Block {
+  hash: Bytes!
+  number: BigInt!
+}
+
+type Error {
+	message: String!
+	
+	# Context for the error.
+	network: String!
+	block: Block!
+	transaction: Bytes
```

Comment:
Our error messages should already include the transaction. I'd prefer to code as little structural information about the underlying network in these errors as possible to keep us flexible. (Even if transactions will probably be a thing everywhere.)

Reply:
Removed.